modules/tinymce/src/core/main/ts/mode/Readonly.ts

```diff
@@ -38,23 +39,54 @@ const registerReadOnlyInputBlockers = (editor: Editor): void => {
     }
   });
 
-  editor.on('input', (e) => {
-    if (!e.isComposing && isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+  // Prevent undo level additions during composition
+  editor.on('input compositionend', (e) => {
+    if (isReadOnly(editor)) {
+      // Skip undo level addition during composition
+      if (!e.isComposing) {
+        const undoLevel = editor.undoManager.add();
+        if (Type.isNonNullable(undoLevel)) {
+          editor.undoManager.undo();
+        }
       }
     }
   });
 
-  editor.on('compositionend', () => {
+  // Set up mutation observer to detect and revert unintended changes
+  const observer = new MutationObserver((mutations) => {
     if (isReadOnly(editor)) {
-      const undoLevel = editor.undoManager.add();
-      if (Type.isNonNullable(undoLevel)) {
-        editor.undoManager.undo();
+      let needsRevert = false;
+
+      mutations.forEach((mutation) => {
```

Comment:
Use `Arr.exists` here instead then you don't need to loop the entire thing to flip a state to true also it's a lot cleaner since you don't need any local mutation.